proposals/jsr-305-custom-nullability-qualifiers.md

```diff
@@ -0,0 +1,321 @@
+- `When.NEVER/When.UNKNOWN` doesn't affect type's nullability
+
+```java
+@TypeQualifierNickname
+@Nonnull(when = When.ALWAYS)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MyNonnull {
+}
+
+@TypeQualifierNickname
+@CheckForNull // a nickname to another type qualifier nickname
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MyNullable {
+}
+
+interface A {
+    @MyNullable
+    String foo(@MyNonnull String x);
+}
+```
+
+`A::foo` in the example above should be available in Kotlin as `fun foo(x: 
+String): String?`
+
+
+### Type qualifier default
+[`@TypeQualifierDefault`](https://aalmiray.github.io/jsr-305/apidocs/javax/annotation/meta/TypeQualifierDefault.html)
+allows introducing annotations that when being applied define the default
+nullability within the scope of the annotated element.
+
+To make it work new annotation class must be annotated with 
+`@TypeQualifierDefault` and `@NonNull` or any type qualifier nickname to the 
+latter.
+The single argument for the `@TypeQualifierDefault` defines a set of element 
+types for which the annotation may enhance a type. 
+
+When determining immediate nullability of a type the Kotlin compiler should look
+for nullability annotation on the type itself at first.
+- If there is one, then it should use it
+- Otherwise, nullability is determined by the closest annotation marked as
+`TypeQualifierDefault` and having appropriate applicability in the argument:
+    - `ElementType.METHOD` for return type of methods
+    - `ElementType.PARAMETER` for value parameters
+    - `ElementType.FIELD` for fields
+    - `ElementType.TYPE_USE` for any top-level type
+    
+Note, that `javax.annotation.ParametersAreNonnullByDefault` annotation should 
+work automatically
+by applying the rules above.
+
+Also, it's important to support reading Java package annotations because it's
+supposed to be the most common way to declare a default nullability.
+
+```java
+@Nonnull
+@TypeQualifierDefault({ElementType.METHOD, ElementType.PARAMETER})
+public @interface NonNullApi {
+}
+
+@Nonnull(when = When.MAYBE)
+@TypeQualifierDefault({ElementType.METHOD, ElementType.PARAMETER})
+public @interface NullableApi {
+}
+```
+```java
+// FILE: test/package-info.java
+
+@NonNullApi // declaring all types in package 'test' as non-nullable by default
+package test;
+
+
+// FILE: test/A.java
+package test;
+
+@NullableApi // overriding default nullability from the package
+interface A {
+    String foo(String x); // fun foo(x: String?): String?
+ 
+    @NotNullApi // overriding default from the class
+    String bar(String x, @Nullable String y); // fun bar(x: String, y: String?): String 
+}
+
+```
+
+### `@Migration` annotation
+Current proposal suggests to add the following meta-annotation in 
+the `kotlin.annotation` package:
+```kotlin
+enum class MigrationStatus {
+    ERROR,
+    WARNING,
+    IGNORE
+}
+
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+annotation class Migration(val status: MigrationStatus)
```

Comment:
Why such a generic name? Is it supposed to be used for something else in the future?

Reply:
If yes, maybe it makes sense to rename `-Xjsr305-annotation-migration` to something like `-Xoverride-migration-status`

Reply:
The name is a subject for discussion as usual, but the flag name you proposed seems to be more appropriate here

Reply:
I mean, what other use cases for this annotation were considered?

If none were, then it probably should reflect that it's about JSR-305 in its name, e.g. `kotlin.jvm.Jsr305Migration`, and the compiler argument should indeed be named something like `-Xjsr305-annotation-migration`.